autogluon/utils/tabular/features/generators/bulk.py

```diff
@@ -0,0 +1,149 @@
+            self._post_generators = []
+        else:
+            post_generators = []
+        self.generators: List[List[AbstractFeatureGenerator]] = pre_generators + generators + post_generators
+
+        for generator_group in self.generators:
+            for generator in generator_group:
+                if not isinstance(generator, AbstractFeatureGenerator):
+                    raise AssertionError(f'generators contains an object which is not an instance of AbstractFeatureGenerator. Invalid generator: {generator}')
+
+        self._feature_metadata_in_unused: FeatureMetadata = None  # FeatureMetadata object based on the original input features that were unused by any feature generator.
+
+    def _fit_transform(self, X: DataFrame, **kwargs) -> (DataFrame, dict):
+        feature_metadata = self.feature_metadata_in
+        for i, generator_group in enumerate(self.generators):
+            self._log(20, f'\tStage {i + 1} Generators:')
+            feature_df_list = []
+            for generator in generator_group:
+                generator.set_log_prefix(log_prefix=self.log_prefix + '\t\t', prepend=True)
+                feature_df_list.append(generator.fit_transform(X, feature_metadata_in=feature_metadata, **kwargs))
+
+            self.generators[i] = [generator for j, generator in enumerate(generator_group) if feature_df_list[j] is not None and len(feature_df_list[j].columns) > 0]
+            feature_df_list = [feature_df for feature_df in feature_df_list if feature_df is not None and len(feature_df.columns) > 0]
+
+            if self.generators[i]:
+                feature_metadata_in_generators = FeatureMetadata.join_metadatas([generator.feature_metadata_in for generator in self.generators[i]], shared_raw_features='error_if_diff')
+            else:
+                feature_metadata_in_generators = FeatureMetadata(type_map_raw=dict())
+
+            if i == 0:  # TODO: Improve this
+                used_features_in = feature_metadata_in_generators.get_features()
+                unused_features_in = [feature for feature in self.feature_metadata_in.get_features() if feature not in used_features_in]
+                self._feature_metadata_in_unused = self.feature_metadata_in.keep_features(features=unused_features_in)
+                self._remove_features_in(features=unused_features_in)
+
+            if self.generators[i]:
+                feature_metadata = FeatureMetadata.join_metadatas([generator.feature_metadata for generator in self.generators[i]], shared_raw_features='error')
+            else:
+                feature_metadata = FeatureMetadata(type_map_raw=dict())
+
+            if not feature_df_list:
+                X = DataFrame(index=X.index)
+            elif len(feature_df_list) == 1:
+                X = feature_df_list[0]
+            else:
+                X = pd.concat(feature_df_list, axis=1, ignore_index=False, copy=False)
```

Comment:
Do generators provide a guarantee unique columns between them (i.e. generator 1 generates col1, col2, generator 2: col2, col3; concat will produce two col2 <- this can create unexpected behaviors downstream)? If not, please consider adding unique prefixes so there is no overlap between different generators.

Reply:
Yes, this is guaranteed from the following code:

```
            if self.generators[i]:
                feature_metadata = FeatureMetadata.join_metadatas([generator.feature_metadata for generator in self.generators[i]], shared_raw_features='error')
            else:
                feature_metadata = FeatureMetadata(type_map_raw=dict())
```

shared_raw_features='error' means an exception will be raised if any output columns share the same name between any generator in a generator group.

Reply:
Regarding unique prefixes, this is planned to be an additional parameter to bulk as an optional, it is currently listed as a TODO at the top of the class.